README.md
# oembed-rs

A generic implementation of [oEmbed](https://oembed.com) 1.0.

[![Latest Version](https://img.shields.io/crates/v/oembed.svg)](https://crates.io/crates/oembed)
[![Documentation](https://docs.rs/oembed/badge.svg)](https://docs.rs/oembed)
[![License](https://img.shields.io/github/license/Flisk/oembed-rs.svg)](LICENSE.txt)

## Currently Unsupported

* Endpoint discovery
* XML responses

## Quick Start

```rust
use std::error::Error;
use std::borrow::Cow;
use oembed::client::*;

struct DummyHttp;

impl Http for DummyHttp {
    fn url_encode<'a>(&mut self, s: &'a str) -> HttpResult<Cow<'a, str>> {
        Ok(s.into())
    }

    fn get(&mut self, _url: &str) -> HttpResult<String> {
        Ok("{
            \"version\": \"1.0\",
            \"type\": \"photo\",
            \"width\": 240,
            \"height\": 160,
            \"title\": \"ZB8T0193\",
            \"url\": \"http://farm4.static.flickr.com/3123/2341623661_7c99f48bbf_m.jpg\",
            \"author_name\": \"Bees\",
            \"author_url\": \"http://www.flickr.com/photos/bees/\",
            \"provider_name\": \"Flickr\",
            \"provider_url\": \"http://www.flickr.com/\"
        }".to_string())
    }
}

let schema = Schema::load_included();
let some_url = "http://www.flickr.com/photos/bees/2341623661/";
let mut http = DummyHttp {};

let response = schema.fetch(&mut http, some_url)
    .expect("Missing provider")
    .expect("Failed to fetch server response");

println!("{:?}", response);
```

## License

The `oembed-rs` crate is under the MIT license, which is included in
[`LICENSE.txt`].

The included provider list ([`src/providers.json`]) is under the MIT
license, which is included in [`LICENSE.providers.txt`].

[`LICENSE.txt`]: LICENSE.txt
[`LICENSE.providers.txt`]: LICENSE.providers.txt
[`src/providers.json`]: src/providers.json


CHANGELOG.md
# oembed-rs changelog

## 0.1.0

* initial release


src/client.rs
use crate::{Endpoint, Error, Provider, Response, Result};
use std::borrow::Cow;

/// Result type for the [`Http`] trait
pub type HttpResult<T> = std::result::Result<T, Box<dyn std::error::Error>>;

/// Defines HTTP operations required by client features
///
/// An implementation of this trait is required for any client operationAn implementation
/// of this trait is required for anything involving HTTP.
pub trait Http {
    /// URL-encode a string so it can be used safely as part of a URL.
    fn url_encode<'a>(&mut self, s: &'a str) -> HttpResult<Cow<'a, str>>;

    /// Retrieve the body of a resource located at `url`.
    fn get(&mut self, url: &str) -> HttpResult<String>;
}

/// Schema containing known oEmbed providers and their endpoints
///
/// The list of providers is currently quite small (~400 elements). For this reason, they
/// are stored in a standard [`Vec`] and looked up with a linear scan.
#[derive(Clone, PartialEq, PartialOrd, Hash, Debug, Default)]
pub struct Schema {
    providers: Vec<Provider>,
}

/// Result of an endpoint search
#[derive(Clone, PartialEq, PartialOrd, Hash, Debug)]
pub struct MatchedEndpoint<'a> {
    /// The provider that the matched endpoint belongs to
    pub provider: &'a Provider,

    /// The matched endpoint
    pub endpoint: &'a Endpoint,

    /// The URL scheme that caused this match
    pub matched_scheme: &'a str,
}

impl Schema {
    /// Load providers from the included schema file
    ///
    /// Schema data is included at build time, so your copy may be out of date. If an
    /// up-to-date schema is required, use [`fetch_latest`][1] or [`fetch_from_url`][2]
    /// instead.
    ///
    /// # Panics
    ///
    /// Panics if the included provider file can't be parsed, which should never happen in
    /// practice.
    ///
    /// [1]: Schema::fetch_latest
    /// [2]: Schema::fetch_from_url
    pub fn load_included() -> Self {
        let json = include_str!("providers.json");
        let providers = serde_json::from_str(&json)
            .expect("Failed to load providers.json. This build of oembed is broken!");

        Self { providers }
    }

    /// Load schema from the public list provided at `https://oembed.com/providers.json`
    pub fn fetch_latest(http: &mut impl Http) -> Result<Self> {
        Self::fetch_from_url(http, "https://oembed.com/providers.json")
    }

    /// Load schema from a specific URL
    pub fn fetch_from_url(http: &mut impl Http, url: &str) -> Result<Self> {
        let s = http.get(url).map_err(|e| Error::HttpGet(e.into()))?;

        let providers = serde_json::from_str(&s).map_err(|e| Error::ParseError(e))?;

        Ok(Self { providers })
    }

    /// Search for the first [`Endpoint`] with a scheme matching `url`
    pub fn match_endpoint(&self, url: &str) -> Option<MatchedEndpoint> {
        for provider in &self.providers {
            for endpoint in &provider.endpoints {
                if let Some(matched_scheme) = endpoint.match_url_scheme(url) {
                    return Some(MatchedEndpoint {
                        provider,
                        endpoint,
                        matched_scheme,
                    });
                }
            }
        }

        None
    }

    /// Fetch an oEmbed response for `url`
    ///
    /// Returns `None` if no endpoint with a scheme matching `url` is found.
    pub fn fetch(&self, http: &mut impl Http, url: &str) -> Option<Result<Response>> {
        self.match_endpoint(url)
            .map(|m| m.endpoint.fetch(http, url))
    }
}

impl Endpoint {
    /// Fetch an oEmbed response for `url` from this endpoint
    pub fn fetch(&self, http: &mut impl Http, url: &str) -> Result<Response> {
        let encoded_url = http.url_encode(url).map_err(|e| Error::HttpUrlEncode(e))?;
        let request_url = format!("{}?format=json&url={}", self.url, encoded_url);

        let s = http.get(&request_url).map_err(|e| Error::HttpGet(e))?;

        serde_json::from_str(&s).map_err(|e| Error::ParseError(e))
    }

    fn match_url_scheme(&self, url: &str) -> Option<&str> {
        self.schemes.as_ref().and_then(|schemes| {
            schemes
                .iter()
                .filter(|s| url_matches_scheme(url, &s))
                .next()
                .map(|s| s.as_str())
        })
    }
}

fn url_matches_scheme(url: &str, scheme: &str) -> bool {
    let mut url_chars = url.chars().peekable();
    let mut scheme_chars = scheme.chars().peekable();

    let mut current_scheme_char = scheme_chars.next();

    while let Some(url_char) = url_chars.next() {
        match current_scheme_char {
            Some('*') if scheme_chars.peek() == url_chars.peek() => (),
            Some('*') => continue,

            Some(scheme_char) if scheme_char == url_char => (),

            Some(_) | None => return false,
        }

        current_scheme_char = scheme_chars.next();
    }

    true
}

#[cfg(test)]
mod tests {
    #[test]
    fn url_matches_scheme() {
        assert_eq!(super::url_matches_scheme("spotify:abc", "spotify:*"), true);

        assert_eq!(
            super::url_matches_scheme("https://youtu.be/v/5mMOsl8qpfc", "https://youtu.be/*"),
            true
        );

        assert_eq!(
            super::url_matches_scheme(
                "https://www.youtube.com/watch?v=5mMOsl8qpfc",
                "https://www.youtube.com/watch*"
            ),
            true
        );

        assert_eq!(
            super::url_matches_scheme(
                "http://www.23hq.com/something/photo/a.jpg",
                "http://www.23hq.com/*/photo/*",
            ),
            true
        )
    }
}


src/lib.rs
//! This crate provides a simple generic implementation of the [oEmbed
//! specification][1] version 1.0.
//!
//! ## Notes and Caveats
//!
//! * No HTTP client mechanism is included; users of this library must provide an
//!   implementation of [`client::Http`] to functions requiring remote resources.
//! * [Discovery](https://oembed.com/#section4) is not currently supported.
//! * XML responses are not currently supported.
//! * *Some endpoints* — not naming names — will return data that doesn't conform with the
//!   specification; such data can't currently be parsed by this library. No decision
//!   on how to address this has been made yet, and suggestions are welcome.
//!
//! # Examples
//!
//! ```
//! use std::error::Error;
//! use std::borrow::Cow;
//! use oembed::client::*;
//!
//! struct DummyHttp;
//!
//! impl Http for DummyHttp {
//!     fn url_encode<'a>(&mut self, s: &'a str) -> HttpResult<Cow<'a, str>> {
//!         Ok(s.into())
//!     }
//!
//!     fn get(&mut self, _url: &str) -> HttpResult<String> {
//!         Ok("{
//!             \"version\": \"1.0\",
//!             \"type\": \"photo\",
//!             \"width\": 240,
//!             \"height\": 160,
//!             \"title\": \"ZB8T0193\",
//!             \"url\": \"http://farm4.static.flickr.com/3123/2341623661_7c99f48bbf_m.jpg\",
//!             \"author_name\": \"Bees\",
//!             \"author_url\": \"http://www.flickr.com/photos/bees/\",
//!             \"provider_name\": \"Flickr\",
//!             \"provider_url\": \"http://www.flickr.com/\"
//!         }".to_string())
//!     }
//! }
//!
//! let schema = Schema::load_included();
//! let some_url = "http://www.flickr.com/photos/bees/2341623661/";
//! let mut http = DummyHttp {};
//!
//! let response = schema.fetch(&mut http, some_url)
//!     .expect("Missing provider")
//!     .expect("Failed to fetch server response");
//!
//! assert_eq!("ZB8T0193", response.title.unwrap());
//! ```
//!
//! [1]: https://oembed.com/

#[macro_use]
extern crate serde;
extern crate serde_json;

pub mod client;

/// Crate-wide error type
#[derive(Debug)]
pub enum Error {
    /// Returned if [`client::Http::url_encode`] failed.
    HttpUrlEncode(Box<dyn std::error::Error>),

    /// Returned if [`client::Http::get`] failed.
    HttpGet(Box<dyn std::error::Error>),

    /// Returned if parsing a response failed.
    ParseError(serde_json::Error),
}

impl std::fmt::Display for Error {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        write!(f, "{:?}", self)
    }
}

type Result<T> = std::result::Result<T, Error>;

/// oEmbed provider
#[derive(Deserialize, Serialize, Clone, PartialEq, PartialOrd, Hash, Debug, Default)]
pub struct Provider {
    #[serde(rename = "provider_name")]
    pub name: String,

    #[serde(rename = "provider_url")]
    pub url: String,

    pub endpoints: Vec<Endpoint>,
}

/// oEmbed endpoint
#[derive(Deserialize, Serialize, Clone, PartialEq, PartialOrd, Hash, Debug, Default)]
pub struct Endpoint {
    pub url: String,
    pub schemes: Option<Vec<String>>,
    pub formats: Option<Vec<String>>,

    /// Not currently supported
    pub discovery: Option<bool>,
}

/// Endpoint response to an oEmbed request
///
/// See section 2.3.4 of the [oEmbed specification][1].
///
/// [1]: https://oembed.com/
#[derive(Deserialize, Serialize, Clone, PartialEq, PartialOrd, Hash, Debug)]
pub struct Response {
    #[serde(flatten, rename(deserialize = "type"))]
    pub response_type: ResponseType,
    pub version: String,
    pub title: Option<String>,
    pub author_name: Option<String>,
    pub author_url: Option<String>,
    pub provider_name: Option<String>,
    pub provider_url: Option<String>,
    pub cache_age: Option<String>,
    pub thumbnail_url: Option<String>,
    pub thumbnail_width: Option<i32>,
    pub thumbnail_height: Option<i32>,
}

/// Type-specific oEmbed response data
///
/// See section 2.3.4 of the [oEmbed specification][1].
///
/// [1]: https://oembed.com/
#[derive(Deserialize, Serialize, Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug)]
#[serde(tag = "type", rename_all(deserialize = "lowercase"))]
pub enum ResponseType {
    Photo {
        url: String,
        width: Option<i32>,
        height: Option<i32>,
    },

    Video {
        html: String,
        width: Option<i32>,
        height: Option<i32>,
    },

    Rich {
        html: String,
        width: Option<i32>,
        height: Option<i32>,
    },

    Link,
}
